### PR TITLE
add support for writing files to containers from cloud-init config

### PR DESCRIPTION
**Pull request by europaul**
We are extending the support for the cloud-init config for containers. While the old config format with
```
ENV1=value1
ENV2=value2
```
still remains available we also introduce the support for the original cloud-init config syntax with support for fields `runcmd` (only for defining envs) and `write_files`, like in the following example
```
#cloud-config
runcmd:
  - ENV1=value1
  - ENV2=value2

write_files:
 - path: /etc/injected_file.txt
   owner: root:root
   permissions: '0644'
   encoding: b64
   content: YmxhYmxh
```
The section `runcmd` will only be used for providing envs, while the new section `write_files` can be used to write files to the containers in the usual cloud-init way. The option to change the owner is however not supported at the moment.

The rest of the field of the cloud-init configuration will be ignored.

### CODE REVIEW

File: pkg/pillar/cmd/domainmgr/domainmgr.go
Comment (eriknordmark) [inline pkg/pillar/cmd/domainmgr/domainmgr.go:1539]: Are these developer debug leftovers?

File: pkg/pillar/cmd/domainmgr/domainmgr.go
Comment (eriknordmark) [inline pkg/pillar/cmd/domainmgr/domainmgr.go:1418]: Can an attacker have file.Path include ".." to escape outside of the rootPath?
Using filepath.Clean on file.Path and then checking if the result starts with "
.." would capture such attempts.

File: pkg/pillar/cmd/domainmgr/domainmgr.go
Comment (eriknordmark) [inline pkg/pillar/cmd/domainmgr/domainmgr.go:1543]: How does this use of meta-data relate to the meta-data content created by EVE/domainmgr and placed in the iso image?

File: pkg/pillar/go.mod
Comment (eriknordmark) [inline pkg/pillar/go.mod:75]: I'd much prefer the origin of that fork (https://github.com/go-yaml/yaml) since it has an MIT+Apache license, but I guess this is brought in by github.com/elotl/cloud-init

**Review by eriknordmark — CHANGES_REQUESTED** [review]
We do need to support the old format since the users have existing cloud-init configured for their edge applications. 
So need to detect which format is in use to handle both.

File: pkg/pillar/go.mod
Comment (europaul) [reply pkg/pillar/go.mod:75]: yes, it's a constraint specified [here](https://github.com/elotl/cloud-init/blob/bf3035f6997c8693f050c40366f9eddbfe0b0ceb/Gopkg.toml#L36C1-L38C34). the implementation is however very simple, so we can do our own if needed

File: pkg/pillar/cmd/domainmgr/domainmgr.go
Comment (europaul) [reply pkg/pillar/cmd/domainmgr/domainmgr.go:1418]: excellent catch!

File: pkg/pillar/cmd/domainmgr/domainmgr.go
Comment (europaul) [reply pkg/pillar/cmd/domainmgr/domainmgr.go:1539]: yes, sorry :smile:

File: pkg/pillar/cmd/domainmgr/domainmgr.go
Comment (europaul) [reply pkg/pillar/cmd/domainmgr/domainmgr.go:1543]: in the case of ISO image we are using the original cloud-init daemon with the data source [NoCloud](https://cloudinit.readthedocs.io/en/latest/reference/datasources/nocloud.html#). it expects a `meta-data` file filled with metadata, like you would find it in the [EC2 metadata service](https://cloudinit.readthedocs.io/en/latest/reference/datasources/ec2.html#amazon-ec2). here we can define a whole number of metadata fields:
```
ami-id
ami-launch-index
ami-manifest-path
block-device-mapping/
hostname
instance-id
instance-type
local-hostname
local-ipv4
placement/
public-hostname
public-ipv4
public-keys/
reservation-id
security-groups
```
Of those fields the `meta-data` file on our VMs only use `instance-id` and `local-hostname` of which the `instance-id` is especially important, since it contains the version of the cloud-init config.

In the case of container as edge-app, we don't rely on a cloud-init daemon, but parse the config manually. Thus of all metadata fields, only cloud-init version remains necessary - this is why I decided to include only the field `instance-id` in its original format in the `meta-data` file.

**Comment by europaul** [comment]
brought back the support for old cloud-init format and fixed unit tests.
I also intend to add integration tests in eden to test the new functionality

File: pkg/pillar/go.mod
Comment (eriknordmark) [reply pkg/pillar/go.mod:75]: I think that would be better. Some users are sensitive to additional *GPL licenses (outside of the common ones like the basic Linux kernel, user space, and Xen).

**Comment by eriknordmark** [comment]
You can ignore those yetus complaints. It is supposed to only look at lines/functions which were changed in your PR but that doesn't always work.

File: pkg/pillar/go.mod
Comment (europaul) [reply pkg/pillar/go.mod]: done

File: pkg/pillar/cmd/domainmgr/domainmgr.go
Comment (eriknordmark) [reply pkg/pillar/cmd/domainmgr/domainmgr.go:1512]: OK

File: pkg/pillar/cmd/domainmgr/domainmgr.go
Comment (eriknordmark) [inline pkg/pillar/cmd/domainmgr/domainmgr.go:1514]: Are we going to log an error the first time we boot a container? I don't see how an initial meta-data file is written before we try to activate aka boot it.

File: pkg/pillar/cmd/domainmgr/domainmgr.go
Comment (eriknordmark) [inline pkg/pillar/cmd/domainmgr/domainmgr.go:1529]: What do you want to write when getCloudInitVersion() failed?

File: pkg/pillar/utils/cloudconfig/cloudconfig.go
Comment (eriknordmark) [inline pkg/pillar/utils/cloudconfig/cloudconfig.go:17]: In https://github.com/lf-edge/eve/pull/3520/files you'll see yetus annotations once yetus has run. Please address the ones which are about in new/changed lines.

File: pkg/pillar/utils/cloudconfig/cloudconfig.go
Comment (eriknordmark) [inline pkg/pillar/utils/cloudconfig/cloudconfig.go:59]: I think it would be better to log file.Path and maybe have the message be that the path tries to escape the root.

File: pkg/pillar/utils/cloudconfig/cloudconfig.go
Comment (eriknordmark) [inline pkg/pillar/utils/cloudconfig/cloudconfig.go:24]: Should we document and enforce any size limits on content?
FWIW I think the controller(s) might have a large limit like 1MB, but maybe we want something smaller here.

**Review by eriknordmark — APPROVED** [review]
Still some minor issues, but approving so we can get some test results.

File: pkg/pillar/cmd/domainmgr/domainmgr.go
Comment (europaul) [reply pkg/pillar/cmd/domainmgr/domainmgr.go:1514]: yes, is this problematic? maybe this should be a warning instead of an error, since it's not critical and we have a fallback in place

Comment (eriknordmark) [reply pkg/pillar/cmd/domainmgr/domainmgr.go:1514]: If it is normal that the file dosn't exist it should be neither error nor warnings. We need to get rid of false errors.

File: pkg/pillar/cmd/domainmgr/domainmgr.go
Comment (europaul) [reply pkg/pillar/cmd/domainmgr/domainmgr.go:1529]: getCloudInitVersion never fails, we either write the version of cloud init present in the config, or we write the version of the config

File: pkg/pillar/utils/cloudconfig/cloudconfig.go
Comment (europaul) [reply pkg/pillar/utils/cloudconfig/cloudconfig.go:24]: The only information that I could find on the topic is that some cloud providers (e.g. AWS) restrict the size to [16kB](https://cloudinit.readthedocs.io/en/latest/explanation/format.html#gzip-compressed-content). Since this restriction is done on the cloud side I think that our controller imposed limit of 1MB might be sufficient.

The only concern that I see here is that the files can fill up EVE's storage, but that should be a decisive factor given that container's sizes are usually in tens of MBs

File: pkg/pillar/utils/cloudconfig/cloudconfig.go
Comment (europaul) [reply pkg/pillar/utils/cloudconfig/cloudconfig.go:19]: Done

File: pkg/pillar/cmd/domainmgr/domainmgr.go
Comment (europaul) [reply pkg/pillar/cmd/domainmgr/domainmgr.go]: I removed the error now

**Comment by europaul** [comment]
added tests for the new cloudconfig package

File: pkg/pillar/cmd/domainmgr/domainmgr.go
Comment (eriknordmark) [reply pkg/pillar/cmd/domainmgr/domainmgr.go:1528]: Sorry I misread the code a few lines earlier.

File: pkg/pillar/utils/cloudconfig/cloudconfig.go
Comment (eriknordmark) [reply pkg/pillar/utils/cloudconfig/cloudconfig.go:27]: Ok

**Review by eriknordmark — APPROVED** [review]
Can you update the PR description to capture that we do handle the old format.

Also, is there any file in docs or pkg/pillar/docs which specifies the env=value format? Any one discuss cloud-init or userdata where we should add such specification?
It might make sense to rewrite/expand the first paragraph in docs/ECO-METADATA.md for this.

Feel free to merge this and do that doc update in a separate PR if you'd like.

**Comment by europaul** [comment]
@eriknordmark I updated the PR description, but I don't have the write access to the repo to merge.

I think we should create a new doc file `docs/CLOUD-INIT.md` where we describe how our ECOs (VMs and containers) support cloud-init configuration.

**Comment by eriknordmark** [comment]
> I think we should create a new doc file `docs/CLOUD-INIT.md` where we describe how our ECOs (VMs and containers) support cloud-init configuration.

Makes sense. I'll merge this for now

**Pull request merged**
